File: svelte/compiler.ts
import { compile as compileSvelte } from "svelte/compiler"

type Input = {
  code: string
  path: string
  target: "ssr" | "dom"
  dev: boolean
}

// Capitalized for Go
type Output =
  | {
      JS: string
      CSS: string
    }
  | {
      Error: {
        Path: string
        Name: string
        Message: string
        Stack?: string
      }
    }

// Compile svelte code
export function compile(input: Input): string {
  const { code, path, target, dev } = input
  const svelte = compileSvelte(code, {
    filename: path,
    generate: target,
    hydratable: true,
    format: "esm",
    dev: dev,
    css: dev,
  })
  return JSON.stringify({
    CSS: svelte.css.code,
    JS: svelte.js.code,
  } as Output)
}

File: svelte/compiler.go
package svelte

//go:generate esbuild compiler.ts --format=iife --global-name=bud_svelte --bundle --platform=node --inject:shimssr.ts --external:url --outfile=compiler.js --log-level=warning

import (
	"encoding/json"
	"fmt"

	"gitlab.com/mnm/bud/ldflag"

	_ "embed"

	"gitlab.com/mnm/bud/js"
)

type Input struct {
	VM  js.VM
	Dev bool
}

func New(vm js.VM) *Compiler {
	return &Compiler{vm, ldflag.IsDevelopment()}
}

type Compiler struct {
	VM  js.VM
	Dev bool
}

type SSR struct {
	JS  string
	CSS string
}

// compiler.js is used to compile .svelte files into JS & CSS
//go:embed compiler.js
var compiler string

// Compile server-rendered code
func (c *Compiler) SSR(path string, code []byte) (*SSR, error) {
	expr := fmt.Sprintf(`%s; bud_svelte.compile({ "path": %q, "code": %q, "target": "ssr", "dev": %t })`, compiler, path, code, c.Dev)
	result, err := c.VM.Eval(path, expr)
	if err != nil {
		return nil, err
	}
	out := new(SSR)
	if err := json.Unmarshal([]byte(result), out); err != nil {
		return nil, err
	}
	return out, nil
}

type DOM struct {
	JS  string
	CSS string
}

// Compile DOM code
func (c *Compiler) DOM(path string, code []byte) (*DOM, error) {
	expr := fmt.Sprintf(`%s; bud_svelte.compile({ "path": %q, "code": %q, "target": "dom", "dev": %t })`, compiler, path, code, c.Dev)
	result, err := c.VM.Eval(path, expr)
	if err != nil {
		return nil, err
	}
	out := new(DOM)
	if err := json.Unmarshal([]byte(result), out); err != nil {
		return nil, err
	}
	return out, nil
}
